router.js

```diff
@@ -1,93 +1,501 @@
+			}
+		}else{
+			this.error(colors.red('ERROR: ') + 'Cannot upgrade socket for websockets because the header host does not exist in the routing table!', req.headers)
 		}
-	} else {
-		do404(res);
-	}
-});
-
-server.on('upgrade', function(req, socket, head) {
-	// Check for an entry in the router table
-	if (routerTable[req.headers.host] != null) {
-		var route = routerTable[req.headers.host];
-		if (route.target) {
-			proxy.ws(req, socket, head, route);
-		} else {
-			console.log('Cannot route ' + req.headers.host + ' because config entry is missing "target" property.');
+	}
+
+	/**
+	 * Sends error with given mesage or default onlySecure
+	 * @param {number}		code Code number of http error
+	 * @param {response}	res Response to send fromJSON
+	 * @param {string}		errMsg = Default error message
+	 */
+	doErrorResponse(code, res, errMsg = this.errors[code] || this.errors['other']) {
+		if (!errMsg) {
+			switch (code) {
+				case 404:
+					errMsg = code + ' Not found'
+					break
+				case 403:
+					errMsg = code + ' Forbiden'
+					break
+				case 503:
+					errMsg = code + ' Service unavailable'
+					break
+				default:
+					errMsg = code + ' An error occurred'
+			}
 		}
-	} else {
-		console.log('Cannot upgrade socket for websockets because the header host does not exist in the routing table!', req.headers);
-	}
-});
-
-proxy.on('proxyError', function (err, req, res) {
-	if (routerTable[req.headers.host] != null) {
-		var route = routerTable[req.headers.host];
-		if (route.errorRedirect) {
-			res.writeHead(302, {'Location': route.errorRedirect});
-			res.end();
-		} else {
-			do404(res);
+
+		if(errMsg[0] == '.'){
+			response.writeHead(code, { 'Content-Type': 'text/HTML; charset=utf-8' })
+			fs.createReadStream(filePath).pipe(response)
+		}else{
+			res.writeHead(code, { 'Content-Type': 'text/plain; charset=utf-8' })
+			res.end(errMsg)
 		}
-	} else {
-		do404(res);
-	}
-	
-	return true;
-});
-
-// Load the initial config data
-loadConfigData(function () {
-	// Config data was loaded so... start the server
-	server.listen(serverPort);
-	
-	// Watch the config file for changes
-	fs.watchFile(configFilePath, function (curr, prev) { configFileEvent(curr, prev); });
-});
\ No newline at end of file
+	}
+
+	static createRoute(route) {
+		return Route.fromJSON(route)
+	}
+}
+
+class Route {
+	constructor({
+		enabled = true,
+		target,
+		secure = true,
+		address = 'localhost',
+		port = 8080,
+		ssl = {
+			enable: true,
+			generate: true,
+			onlySecure: true
+		},
+		errorRedirect
+	} = {}){
+		this.enabled = enabled
+		this.target = target || `http${secure?'s':''}://${address}:${port}`
+		this.ssl = ssl
+		this.errorRedirect = errorRedirect
+	}
+
+	toProxy() {
+		return { target: this.target }
+	}
+
+	static fromJSON(object){
+		return new Route(object)
+	}
+}
+
+const router = new Router({
+	email: 'akxe@example.com',
+	routesFile: './exampleConfig.json'
+})
+router.listen()
```

Comment:
I'm thinking that given this is a complete re-write, some tests should be made to ensure this works. Including websocket tests as well as we are supposed to support them too.

Reply:
Definitely agreed, however I am unable to resolve implementation issues with greenlock, thus it is not working with SSL. And second problem is that I don't know how to create tests, if you create one, I can copycat the others.